File: ts/marquee-markup/bin/marquee.ts
#!/usr/bin/env node
// The marquee CLI: one file to a page, or one folder to a website.
//
//     marquee hello.mq > hello.html      a self-contained page on stdout
//     marquee hello.mq -o hello.html     or written to a file
//     marquee site/ dist/                a whole site (shared includes,
//                                        per-site font subsetting)
//
// Batteries included, no surprises: by default the CLI runs the turbolink
// fetch-ahead pass (OpenGraph summaries for bare web links). --nofetch
// produces the spartan, zero-network output instead.

import { readFileSync, statSync, writeFileSync } from "node:fs";
import { buildSite, buildSiteFetch, marquee, marqueeFetch } from "../src/index.ts";

// Piping into `head` is not an emergency.
process.stdout.on("error", (err: NodeJS.ErrnoException) => {
  if (err.code === "EPIPE") {
    process.exit(0);
  }
  throw err;
});

function usage(): never {
  console.error(
    [
      "usage:",
      "  marquee <file.mq> [-o out.html]   render one self-contained page",
      "  marquee <site-dir> <out-dir>      build a whole site",
      "  --nofetch                         skip the fetch-ahead pass (no network,",
      "                                    web turbolinks stay plain links)",
      "  --envelope                        wrap plain documents in a 650px centered",
      "                                    envelope for readability (documents with",
      "                                    their own :::page layout are left alone)",
      "  --darkmode                        force dark mode (default: follow the",
      "                                    reader's OS theme)",
      "  --noreadable                      don't rescue author colors (default: their",
      "                                    lightness is clamped toward the canvas's",
      "                                    opposite so colored text stays legible)",
    ].join("\n"),
  );
  process.exit(2);
}

const args = process.argv.slice(2);
const positional: string[] = [];
let outFile: string | null = null;
let fetchMode = true;
let envelope = false;
let readable = true;
let colorScheme: "light" | "dark" | undefined;
for (let i = 0; i < args.length; i += 1) {
  if (args[i] === "-o") {
    outFile = args[++i] ?? null;
    if (outFile === null) {
      usage();
    }
    continue;
  }
  if (args[i] === "--nofetch") {
    fetchMode = false;
    continue;
  }
  if (args[i] === "--envelope") {
    envelope = true;
    continue;
  }
  if (args[i] === "--darkmode") {
    colorScheme = "dark";
    continue;
  }
  if (args[i] === "--noreadable") {
    readable = false;
    continue;
  }
  positional.push(args[i]!);
}
if (positional.length === 0 || positional.length > 2) {
  usage();
}

const opts = { envelope, readable, ...(colorScheme === undefined ? {} : { colorScheme }) };

const [input, outDir] = positional;
const isDir = (() => {
  try {
    return statSync(input!).isDirectory();
  } catch {
    console.error(`marquee: ${input}: not found`);
    process.exit(1);
  }
})();

if (isDir) {
  if (outDir === undefined) {
    usage();
  }
  const report = fetchMode
    ? await buildSiteFetch(input!, outDir, opts)
    : buildSite(input!, outDir, opts);
  console.error(
    `built ${report.pages.length} pages (${report.pages.join(", ")}) + ${report.mediaFiles} media files + ${report.fontFaces.length} font faces -> ${report.outDir}`,
  );
} else {
  if (outDir !== undefined) {
    usage();
  }
  const source = readFileSync(input!, "utf8");
  const page = fetchMode ? await marqueeFetch(source, opts) : marquee(source, opts);
  if (outFile === null) {
    process.stdout.write(page);
  } else {
    writeFileSync(outFile, page);
    console.error(`wrote ${outFile}`);
  }
}

File: rust/markup/src/bin/marquee.rs
//! The marquee CLI, Rust spelling - flag-for-flag with the npm one:
//!
//!     marquee hello.mq > hello.html      a self-contained page on stdout
//!     marquee hello.mq -o hello.html     or written to a file
//!     marquee site/ dist/                a whole site
//!
//! Batteries included, no surprises: by default the CLI runs the turbolink
//! fetch-ahead pass (OpenGraph summaries for bare web links). --nofetch
//! produces the spartan, zero-network output instead.

use marquee_markup::{
    build_site, build_site_fetch, marquee, marquee_fetch, ColorScheme, MarqueeOptions,
    SiteOptions,
};
use std::path::Path;
use std::process::exit;

fn usage() -> ! {
    eprintln!(
        "usage:
  marquee <file.mq> [-o out.html]   render one self-contained page
  marquee <site-dir> <out-dir>      build a whole site
  --nofetch                         skip the fetch-ahead pass (no network,
                                    web turbolinks stay plain links)
  --envelope                        wrap plain documents in a 650px centered
                                    envelope for readability (documents with
                                    their own :::page layout are left alone)
  --darkmode                        force dark mode (default: follow the
                                    reader's OS theme)
  --noreadable                      don't rescue author colors (default: their
                                    lightness is clamped toward the canvas's
                                    opposite so colored text stays legible)"
    );
    exit(2);
}

fn main() {
    let args: Vec<String> = std::env::args().skip(1).collect();
    let mut positional: Vec<String> = Vec::new();
    let mut out_file: Option<String> = None;
    let mut fetch_mode = true;
    let mut envelope = false;
    let mut readable: Option<bool> = None;
    let mut color_scheme: Option<ColorScheme> = None;
    let mut i = 0;
    while i < args.len() {
        match args[i].as_str() {
            "-o" => {
                i += 1;
                out_file = Some(args.get(i).cloned().unwrap_or_else(|| usage()));
            }
            "--nofetch" => fetch_mode = false,
            "--envelope" => envelope = true,
            "--darkmode" => color_scheme = Some(ColorScheme::Dark),
            "--noreadable" => readable = Some(false),
            other => positional.push(other.to_string()),
        }
        i += 1;
    }
    if positional.is_empty() || positional.len() > 2 {
        usage();
    }
    let input = Path::new(&positional[0]);
    if !input.exists() {
        eprintln!("marquee: {}: not found", positional[0]);
        exit(1);
    }

    if input.is_dir() {
        let Some(out_dir) = positional.get(1) else { usage() };
        let opts = SiteOptions {
            envelope,
            readable,
            color_scheme,
            ..SiteOptions::default()
        };
        let result = if fetch_mode {
            build_site_fetch(input, Path::new(out_dir), &opts)
        } else {
            build_site(input, Path::new(out_dir), &opts)
        };
        match result {
            Ok(report) => eprintln!(
                "built {} pages ({}) + {} media files + {} font faces -> {}",
                report.pages.len(),
                report.pages.join(", "),
                report.media_files,
                report.font_faces.len(),
                report.out_dir.display()
            ),
            Err(e) => {
                eprintln!("marquee: {e}");
                exit(1);
            }
        }
    } else {
        if positional.len() != 1 {
            usage();
        }
        let source = match std::fs::read_to_string(input) {
            Ok(s) => s,
            Err(e) => {
                eprintln!("marquee: {}: {e}", positional[0]);
                exit(1);
            }
        };
        let opts = MarqueeOptions {
            envelope,
            readable,
            color_scheme,
            ..MarqueeOptions::default()
        };
        let page = if fetch_mode {
            marquee_fetch(&source, &opts)
        } else {
            marquee(&source, &opts)
        };
        match (page, out_file) {
            (Ok(page), None) => print!("{page}"),
            (Ok(page), Some(path)) => match std::fs::write(&path, page) {
                Ok(()) => eprintln!("wrote {path}"),
                Err(e) => {
                    eprintln!("marquee: {path}: {e}");
                    exit(1);
                }
            },
            (Err(e), _) => {
                eprintln!("marquee: {e}");
                exit(1);
            }
        }
    }
}
